### ISSUE TITLE
No Generator is considered to be an instance of GeneratorType

### ISSUE THREAD

**Issue by mthuurne** [python/mypy#10501]
**Bug Report**

`types.GeneratorType` is [documented as](https://docs.python.org/3/library/types.html#types.GeneratorType) "The type of generator-iterator objects, created by generator functions." Therefore I would expect that a value annotated using `typing.Generator` would be considered an instance of `GeneratorType`. However, that is not the case.

It is possible that this issue originates in typeshed instead of mypy; I don't know whether mypy is drawing the wrong conclusions or is operating on flawed definitions. Even if the problem is in typeshed, perhaps mypy can output a more detailed error message, listing which method signatures it considers incompatible.

**To Reproduce**

Run mypy on the following code:

```py
from types import GeneratorType
from typing import Any, Generator

def check(g: Generator[Any, Any, Any]) -> None:
    if not isinstance(g, GeneratorType):
        raise TypeError("Not a generator")
    print("It is a generator")

def gen() -> Generator[int, None, None]:
    yield 123

check(gen())

try:
    check(None)
except TypeError:
    print("Non-generator is rejected")
else:
    print("Non-generator is accepted")
```

**Expected Behavior**

Only `check(None)` should be flagged as an error.

**Actual Behavior**

```
$ mypy generator_type.py 
generator_type.py:5: error: Subclass of "Generator[Any, Any, Any]" and "GeneratorType" cannot exist: would have incompatible method signatures  [unreachable]
generator_type.py:7: error: Statement is unreachable  [unreachable]
generator_type.py:15: error: Argument 1 to "check" has incompatible type "None"; expected "Generator[Any, Any, Any]"  [arg-type]
Found 3 errors in 1 file (checked 1 source file)
```

The first error (`isinstance`), reported on line 5, is the one I consider a bug.
The error reported on line 7 (`print`) is a consequence of the one on line 5.
The error reported on line 15 (`check(None)`) is an actual error.

Note that the runtime behavior is as expected:
```
$ python generator_type.py 
It is a generator
Non-generator is rejected
```

**Your Environment**

I'm using mypy 0.812 with Python 3.8.8 on openSUSE Linux.

**Comment by JelleZijlstra** [comment]
You're in luck because I just submitted a typeshed PR that should fix this: python/typeshed#5474.

**Comment by mthuurne** [comment]
Thanks!

I'm trying to verify that your PR fixes this, but I can't get mypy to use an external typeshed directory:
```
$ mypy --custom-typeshed-dir ~/src/typeshed generator_type.py
generator_type.py:1: error: No library stub file for standard library module 'types'  [import]
generator_type.py:1: note: (Stub files are from https://github.com/python/typeshed)
generator_type.py:1: error: Cannot find 'builtins' module. Typeshed appears broken!
generator_type.py:2: error: No library stub file for standard library module 'typing'  [import]
Found 3 errors in 1 file (errors prevented further checking)
```
Where `~/src/typeshed` is the git work area where I checked out your branch.

**Comment by JelleZijlstra** [comment]
You may have to also use mypy master. mypy hasn't had a release yet since we changed the typeshed directory structure a few months ago.

**Comment by mthuurne** [comment]
Indeed it works with mypy master and the unexpected errors are gone then.

Should I close this issue now or wait until the fix has been merged?

**Comment by JelleZijlstra** [comment]
I'll transfer the issue to typeshed and associate the PR with the issue.

**Comment by JelleZijlstra** [comment]
Also, thanks for checking!
